samples/Matty/Matty.Server/Startup.cs

```diff
@@ -0,0 +1,171 @@
+                // register an email sender service (IEmailSender) and
+                // set options.SignIn.RequireConfirmedAccount to true.
+                //
+                // For more information, visit https://aka.ms/aspaccountconf.
+                options.SignIn.RequireConfirmedAccount = false;
+            });
+
+            // OpenIddict offers native integration with Quartz.NET to perform scheduled tasks
+            // (like pruning orphaned authorizations/tokens from the database) at regular intervals.
+            services.AddQuartz(options =>
+            {
+                options.UseMicrosoftDependencyInjectionJobFactory();
+                options.UseSimpleTypeLoader();
+                options.UseInMemoryStore();
+            });
+
+            // Register the Quartz.NET service and configure it to block shutdown until jobs are complete.
+            services.AddQuartzHostedService(options => options.WaitForJobsToComplete = true);
+
+            services.AddOpenIddict()
+
+                // Register the OpenIddict core components.
+                .AddCore(options =>
+                {
+                    // Configure OpenIddict to use the Entity Framework Core stores and models.
+                    // Note: call ReplaceDefaultEntities() to replace the default OpenIddict entities.
+                    options.UseEntityFrameworkCore()
+                           .UseDbContext<ApplicationDbContext>();
+
+                    // Enable Quartz.NET integration.
+                    options.UseQuartz();
+                })
+
+                // Register the OpenIddict server components.
+                .AddServer(options =>
+                {
+                    // Enable the authorization, logout, token and userinfo endpoints.
+                    options.SetTokenEndpointUris("/connect/token")
+                           .SetUserinfoEndpointUris("/connect/userinfo")
+                           .SetDeviceEndpointUris("/connect/device")
+                           .SetVerificationEndpointUris("/connect/verify");
+
+                    // Mark the "email", "profile" and "roles" scopes as supported scopes.
+                    options.RegisterScopes(Scopes.Email, Scopes.Profile, Scopes.Roles);
+
+                    // Note: this sample only uses the authorization code flow but you can enable
+                    // the other flows if you need to support implicit, password or client credentials.
+                    options.AllowDeviceCodeFlow();
+                    options.AllowRefreshTokenFlow();
+
+                    // Note: All device codes could be exchanged once with the tokens.
+                    // Using short-lived device codes is strongly recommended.
+                    // The default Lifetime is null which can be specified to issue codes
+                    // that never expire before exchanged with the tokens
+                    options.SetDeviceCodeLifetime(TimeSpan.FromMinutes(5));
```

Comment:
I don't think that's true: device codes, like user codes, have a default lifetime of 10 minutes, after which they can't be used anymore. You may have been mistaken because of 2 things:
  - The expiration date check is made after the token entry status check, so you don't an "expired device code" error until the user accepts or rejects the authorization demand.
  - When the user approves the authorization demand, the expiration associated with the device code entry is extended to give plenty of time to the client to redeem the device code if the user approved the demand at the very last moment (i.e just before the 10mn limit).

These are 2 aspects I plan to change in 4.0: the expiration check will be made earlier and I'll likely remove the "dynamic expiration date" update mechanism to simply configure device codes to use a longer lifetime than user codes by default.

Reply:
The method name does not indicate its usage, neither does the description.
> Sets the device code lifetime, after which client applications are unable to
> send a grant_type=urn:ietf:params:oauth:grant-type:device_code token request.
>Using short-lived device codes is strongly recommended. While discouraged, null
> can be specified to issue codes that never expire.

Does this note seem correct?
```
// Note: All device codes have a default of 10 minutes.
// All device codes could be exchanged once with the tokens.
// This option is used to extend the lifetime of the device code
// after the user approves the authorization demand to give
// the client some time to exchange the device code with the token
// when the user approval is granted at the very last moment
// (i.e just before the 10 minutes limit).
```

Does `SetUserCodeLifetime` use a similar behavior?

Reply:
I don't think any of this is necessary. Consider removing that and leaving the default value (10 minutes).